doc/modules/classes.rst

```diff
@@ -1058,9 +1058,11 @@ Hyper-parameter optimizers
    :template: class.rst
 
    model_selection.GridSearchCV
+   model_selection.GridSuccessiveHalving
```

Comment:
the lack of CV or even SearchCV at the end may be controversial!

Reply:
I agree. There should probably be a CV. I'm not sure what a better name would be, though. ``GridSHSearchCV`` is obscure ``GridHalvingSearchCV``?

Reply:
why completely new class and not just adding an early stop parameter with default off?
I think that it may be part of both Grid and Random search... For me, the early stopping option really makes sense especially in case of RandomizedSearchCV when you may hit the best configuration almost any time and not wasting your resources... :)

Reply:
I don't see the relation between SH and early stopping?

Reply:
@NicolasHug what is "SH"?

Reply:
SH = Successive halving.

What do you mean by early stopping? The estimator fit or the search? Can you please elaborate on your proposal?

Reply:
in fact, it can be in both, but first I had in my mind early stopping for the search... 
As you have a case where you are fine with reaching accuracy 95% then and accidentally you get this result after the first iteration of CV, there is no need to continue the search. So I would implement it as adding one more parameter to the {Random, Grid}SearchCV meaning you break the loop if performance gets better then some threshold. If not specified, simply run the original search. Does it make sense to you?

Reply:
@Borda that kind of makes sense to me. Do you think it's common to know what your threshold is, though? We could have a stopping criterion like "n_iter_no_improvement" that is more adaptive, and that's actually quite common in Bayesian Optimization. If this is optional that would be fine.
One tricky aspect is that this would interact with parallelization somewhat.

Feel free to open a PR or issue on this. For early stopping the estimators, we have #8230 but that is arguably more tricky (well and we have the EstimatorCV models).

Your proposal makes sense but is unrelated to successive halving so I don't think we should discuss it here.

Reply:
Successive halving consists in trying many parameter combinations with a low budget, keep the best half, and repeat with a higher budget (keeping the best halg again, etc).

If you get a high accuracy at the first few iterations for a given parameter combination, it might just be because you got lucky and you just overfit a small training set (budget = train set size). It doesn't mean you should select it as the best one right away: it might be outperformed by other combinations when using a higher budget.

Hope that makes sense.

Reply:
@amueller the case with `n_iter_no_improvement` sounds interesting but it comes to the "problem" from completely another perspective, imagine that you hit local minima in the first iteration and then you on a flat ground for a while... this raises that the first minima is enough even it may be the `n_iter_no_improvement +1` worse result you can get in case of grid search. For a random search, it may be a bit better but still, it would be quite a week.
I do agree that the defining the threshold for early sop require some a prior knowledge of the  domain/problem or to have some expectation, but recently I was facing question from industry "Give me a classifier with X% accuracy as soon as possible so we can continue decision..." So what you can do then (assuming RandomSearchCV):
1. run an exhaustive search which may take ages
2. run a small search and if your results are not good enough try it again with a larger number of iteration
the disadvantage of 1) is long execution time and 2) user interaction; compare to early stopping I would have a result soon if it exists...
I was asking here #1626, so shall I continue the discussion there or open a new issue (I would not start a PR until is clear that such feature is welcome)

Reply:
@NicolasHug it sounds like an alternative to genetic algo where you generate a population of parameters and keeps just the nest and in next generation, you make a random change on some subset of parameters, right?

that is quite similar to what we implemented recently for DNN architectures...

Reply:
you don't make any random change for the next generation, just use the survivors of the previous generations. But yes this is, more or less, the same idea.

In any case like Andreas suggested, this is not specific to SH, so we should discuss this in another issue.

Reply:
@borda commented at #1626 as indeed this is a separate issue.